### PR TITLE
Add iot_sdio.h and address PR comments.

### PR DESCRIPTION
**Pull request by qizengqizeng**
Add iot_sdio.h and address PR comments.
Description
-----------

Checklist:
----------

- [x] I have tested my changes. No regression in existing tests.
- [ ] My code is Linted.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: libraries/abstractions/common_io/include/iot_sdio.h [hunk 1]
```diff
@@ -0,0 +1,709 @@
...
+ * @{
+ */
+
+/**
+ * @brief The return codes for the functions in SDIO
+ */
+ /*!< SDIO operation completed successfully. */
+#define IOT_SDIO_SUCCESS                             ( 0 )
+/*!< SDIO operation failed. */
+#define IOT_SDIO_FAIL                                ( 1 )
```
Comment (qiutongs) [inline libraries/abstractions/common_io/include/iot_sdio.h]: If this is specific to ioctl, can we rename it to IOT_SDIO_IOCTL_FAIL

File: libraries/abstractions/common_io/include/iot_sdio.h [hunk 2]
```diff
@@ -0,0 +1,709 @@
...
+ * @file    iot_sdio.h
+ * @brief   This file contains all the SDIO HAL API definitions.
+ *          Here we define SDIO driver as a protocol layer driver which builds on top of SDMMC
+ *          host driver. SDMMC host driver operates at physical layer and provide APIs to send
+ *          commands to the slave device and receive command responses, send and receive data,
+ *          and handle error conditions on the bus.
+ *
+ *          SD/SDIO/MMC driver can be built on top of SDMMC host. This interface only define
+ *          the APIs for SDIO applications, the interface can be expanded to support SD/MMC either
+ *          by expanding the APIs in this file or adding separate files for SD/MMC
```
Comment (dcgaws) [inline libraries/abstractions/common_io/include/iot_sdio.h]: Can you please add a period to the end of the sentence?

File: libraries/abstractions/common_io/include/iot_sdio.h [hunk 3]
```diff
@@ -0,0 +1,709 @@
...
+ *
+ * @design consideration
+ *          SDIO HAL APIs performs protocol layer tasks between sdio host and sdio card.
```
Comment (dcgaws) [inline libraries/abstractions/common_io/include/iot_sdio.h]: SDIO is capitalized above.

File: libraries/abstractions/common_io/include/iot_sdio.h [hunk 4]
```diff
@@ -0,0 +1,709 @@
...
+ *          Some APIs perform a single task and some APIs perform a sequence of tasks.
+ *          For examples, iot_sdio_io_read_direct() API performs a single task of reading
+ *          one byte from card register, and iot_sdio_card_connect() API performs all tasks
+ *          required by card initialization as specified in section 3.1.2 in "SD Specification
+ *          Part E1 SDIO".
+ *
+ *          Upon calling iot_sdio_open() API, the user is a given a pxSdioHandle, an abstract
```
Comment (dcgaws) [inline libraries/abstractions/common_io/include/iot_sdio.h]: Is it correct to call it abstract once it's been instantiated and bound to a specific device?

File: libraries/abstractions/common_io/include/iot_sdio.h [hunk 5]
```diff
@@ -0,0 +1,709 @@
...
+ *          reference to the SDIO card slot that the user is interested in. Thereafter, user
+ *          uses this handle to call other APIs to perform tasks associated with the target
+ *          card slot. For example, iot_sdio_check_card_presence_status(pxSdioHandle, true)
+ *          checks sdio card presence, and iot_sdio_io_write_extended(pxSdioHandle)
+ *          writes data to sdio card in the slot, etc.
+ *
+ * @limitation
+ *          SDIO HAL APIs are intended to be used in embedded systems where each sdio host
+ *          controller is connected to either a single hot swappable sido card slot or multiple
```
Comment (dcgaws) [inline libraries/abstractions/common_io/include/iot_sdio.h]: "sido"

File: libraries/abstractions/common_io/include/iot_sdio.h [hunk 6]
```diff
@@ -0,0 +1,709 @@
...
+/*!< At least one parameter is invalid. */
+#define IOT_SDIO_INVALID_VALUE                       ( 3 )
+/*!< Sending command to or receive response from
+     card failed. */
+#define IOT_SDIO_TRANSFER_FAIL                       ( 4 )
+/*!< If the card is initialized in 3.3V signaling,
+     and card support high speed mode (SHS ==1),
+     host can issue CMD52 in RAW mode, setting EHS
+     bit in CCCR to one to switch to high speed mode
+     . This error code indicates that the sdio card
```
Comment (dcgaws) [inline libraries/abstractions/common_io/include/iot_sdio.h]: Period should be on previous line. More generally, comments could instead be justified to 80 columns.

File: libraries/abstractions/common_io/include/iot_sdio.h [hunk 7]
```diff
@@ -0,0 +1,709 @@
...
+/**
+ * @brief common notify event types in sdio callback
+ */
+typedef enum IotSdioEventType
+{
+    eSdioCardInterruptEvent,    /*!< Interrupt received from sdio card */
+    eSdioCardInsertedEvent,     /*!< sdio card insertion is detected */
+    eSdioCardRemovedEvent,      /*!< sdio card removal is detected */
+    eSdioCardPowerOnEvent,      /*!< Powering on sdio card */
+    eSdioCardPowerOffEvent,     /*!< powering off sdio card */
```
Comment (dcgaws) [inline libraries/abstractions/common_io/include/iot_sdio.h]: Inconsistent sentence case.

File: libraries/abstractions/common_io/include/iot_sdio.h [hunk 8]
```diff
@@ -0,0 +1,709 @@
...
+} IotSdioEventType_t;
+
+/**
+ * @brief Ioctl request types.
```
Comment (dcgaws) [inline libraries/abstractions/common_io/include/iot_sdio.h]: Inconsistent capitalization of IOCTL and/or the @brief comment.

File: libraries/abstractions/common_io/include/iot_sdio.h (hunk 1 above)
Comment (qizengqizeng) [reply libraries/abstractions/common_io/include/iot_sdio.h]: renamed.

File: libraries/abstractions/common_io/include/iot_sdio.h (hunk 2 above)
Comment (qizengqizeng) [reply libraries/abstractions/common_io/include/iot_sdio.h]: added.

File: libraries/abstractions/common_io/include/iot_sdio.h (hunk 3 above)
Comment (qizengqizeng) [reply libraries/abstractions/common_io/include/iot_sdio.h]: capitalized.

File: libraries/abstractions/common_io/include/iot_sdio.h (hunk 4 above)
Comment (qizengqizeng) [reply libraries/abstractions/common_io/include/iot_sdio.h]: removed "abstract"

File: libraries/abstractions/common_io/include/iot_sdio.h (hunk 5 above)
Comment (qizengqizeng) [reply libraries/abstractions/common_io/include/iot_sdio.h]: capitalized

File: libraries/abstractions/common_io/include/iot_sdio.h (hunk 6 above)
Comment (qizengqizeng) [reply libraries/abstractions/common_io/include/iot_sdio.h]: adjusted.

File: libraries/abstractions/common_io/include/iot_sdio.h (hunk 7 above)
Comment (qizengqizeng) [reply libraries/abstractions/common_io/include/iot_sdio.h]: Capitalized first letters.

File: libraries/abstractions/common_io/include/iot_sdio.h (hunk 8 above)
Comment (qizengqizeng) [reply libraries/abstractions/common_io/include/iot_sdio.h]: Capitalized IOCTL

**Pull request merged**